File: http_client/html_parser.py
from utils.constants import SELF_CLOSING_TAGS, HEAD_TAGS, INHERITED_PROPERTIES
import html.entities
class Element:
    def __init__(self, tag, attributes, parent):
        self.tag = tag
        self.attributes = attributes
        self.children = []
        self.parent = parent

    def __repr__(self):
        attrs = [" " + k + "=\"" + v + "\"" for k, v  in self.attributes.items()]
        attr_str = ""
        for attr in attrs:
            attr_str += attr
        return "<" + self.tag + attr_str + ">"

class Text:
    def __init__(self, text, parent):
        self.text = text
        self.children = []
        self.parent = parent

    def __repr__(self):
        return repr(self.text)

class HTML():
    def __init__(self, raw_html):
        self.raw_html = raw_html
        self.unfinished = []

        self.parse()
    
    def parse(self):
        text = ""
        in_tag = False
        for c in self.raw_html:
            if c == "<":
                in_tag = True
                if text: self.add_text(text) # start of new tag means before everything was content/text
                text = ""
            elif c == ">":
                in_tag = False
                self.add_tag(text) # end of a tag means everything in-between were tags
                text = ""
            else:
                text += c
        
        if not in_tag and text:
            self.add_text(text)

        return self.finish()

    def add_text(self, text):
        if text.isspace(): return
        self.implicit_tags(None)
        parent = self.unfinished[-1]
        node = Text(text, parent)
        parent.children.append(node)

    def get_attributes(self, text):
        parts = text.split()
        tag = parts[0].casefold()
        attributes = {}
        
        for attrpair in parts[1:]:
            if "=" in attrpair:
                key, value = attrpair.split("=", 1)
                if len(value) > 2 and value[0] in ["'", "\""]:
                    value = value[1:-1]
                attributes[key.casefold()] = value
            else:
                attributes[attrpair.casefold()] = ""

        return tag, attributes

    def add_tag(self, tag):
        tag, attributes = self.get_attributes(tag)
        
        if tag.startswith("!"): return
        
        self.implicit_tags(tag)

        if tag.startswith("/"):
            if len(self.unfinished) == 1: return
            node = self.unfinished.pop()
            parent = self.unfinished[-1]
            parent.children.append(node)
        elif tag in SELF_CLOSING_TAGS:
            parent = self.unfinished[-1]
            node = Element(tag, attributes, parent)
            parent.children.append(node)
        else:
            parent = self.unfinished[-1] if self.unfinished else None
            node = Element(tag, attributes, parent)
            self.unfinished.append(node)

    def implicit_tags(self, tag):
        while True:
            open_tags = [node.tag for node in self.unfinished]
            if open_tags == [] and tag != "html":
                self.add_tag("html")
            elif open_tags == ["html"] and tag not in ["head", "body", "/html"]:
                if tag in HEAD_TAGS:
                    self.add_tag("head")
                else:
                    self.add_tag("body")
            elif open_tags == ["html", "head"] and tag not in ["/head"] + HEAD_TAGS:
                self.add_tag("/head")
            else:
                break

    def finish(self):
        if not self.unfinished:
            self.implicit_tags(None)
    
        while len(self.unfinished) > 1:
            node = self.unfinished.pop()
            parent = self.unfinished[-1]
            parent.children.append(node)
        return self.unfinished.pop()
    
    @staticmethod
    def print_tree(node, indent=0):
        print(" " * indent, node)
        for child in node.children:
            HTML.print_tree(child, indent + 2)

    @staticmethod
    def to_json(tree: Element | Text):
        if isinstance(tree, Text):
            return ["text", tree.text, [HTML.to_json(child) for child in tree.children]]
        elif isinstance(tree, Element):
            return ["element", tree.tag, tree.attributes, [HTML.to_json(child) for child in tree.children]]

    @staticmethod
    def from_json(json_list, parent=None):
        if json_list[0] == "text":
            text = Text(json_list[1], parent)
            text.children = [HTML.from_json(child, text) for child in json_list[2]]
            return text
        elif json_list[0] == "element":
            element = Element(json_list[1], json_list[2], parent)
            element.children = [HTML.from_json(child, element) for child in json_list[3]]
            return element
        
class TagSelector:
    def __init__(self, tag):
        self.tag = tag
        self.priority = 1
    
    def matches(self, node):
        return isinstance(node, Element) and self.tag == node.tag
    
class DescendantSelector:
    def __init__(self, ancestor, descendant):
        self.ancestor = ancestor
        self.descendant = descendant
        self.priority = ancestor.priority + descendant.priority

    def matches(self, node):
        if not self.descendant.matches(node): return False
        while node.parent:
            if self.ancestor.matches(node.parent): return True
            node = node.parent
        return False

def cascade_priority(rule):
    selector, body = rule
    return selector.priority

def get_inline_styles(node):
    all_rules = []

    for node in node.children:
        if isinstance(node, Element) and node.tag == "style":
            if isinstance(node.children[0], Text):
                all_rules.extend(CSSParser(node.children[0].text).parse()) # node's first children will just be a text element that contains the css

        all_rules.extend(get_inline_styles(node))
    
    return all_rules

class CSSParser:
    def __init__(self, s):
        self.s = s
        self.i = 0

    def whitespace(self):
        while self.i < len(self.s) and self.s[self.i].isspace():
            self.i += 1

    def literal(self, literal):
        if not (self.i < len(self.s) and self.s[self.i] == literal):
            raise Exception("Parsing error")
        self.i += 1

    def word(self):
        start = self.i
        while self.i < len(self.s):
            if self.s[self.i].isalnum() or self.s[self.i] in "#-.%":
                self.i += 1
            else:
                break
        if not (self.i > start):
            raise Exception("Parsing error")
        return self.s[start:self.i]
        
    def pair(self):
        prop = self.word()
        
        self.whitespace()
        self.literal(":")
        self.whitespace()
        
        val = self.word()

        return prop.casefold(), val

    def ignore_until(self, chars):
        while self.i < len(self.s):
            if self.s[self.i] in chars:
                return self.s[self.i]
            else:
                self.i += 1
        return None

    def body(self):
        pairs = {}
        while self.i < len(self.s) and self.s[self.i] != "}":
            try:
                prop, val = self.pair()
                pairs[prop] = val
                
                self.whitespace()
                
                self.literal(";")

                self.whitespace()        
            except Exception:
                why = self.ignore_until([";", "}"])
                if why == ";":
                    self.literal(";")
                    self.whitespace()
                else:
                    break

        return pairs
    
    def selector(self):
        out = TagSelector(self.word().casefold())
        self.whitespace()
        while self.i < len(self.s) and self.s[self.i] != "{":
            tag = self.word()
            descendant = TagSelector(tag.casefold())
            out = DescendantSelector(out, descendant)
            self.whitespace()
        return out
        
    def parse(self):
        rules = []
        while self.i < len(self.s):
            try:
                self.whitespace()
                
                selector = self.selector()
                
                self.literal("{")
                
                self.whitespace()
                
                body = self.body()
                
                self.literal("}")

                rules.append((selector, body))
            except Exception:
                why = self.ignore_until(["}"])
                if why == "}":
                    self.literal("}")
                    self.whitespace()
                else:
                    break
        return rules
    
    @classmethod
    def convert_selector_to_json(self, selector):
        if isinstance(selector, TagSelector):
            return ["tag", selector.tag, selector.priority]
        elif isinstance(selector, DescendantSelector):
            return ["descendant", self.convert_selector_to_json(selector.ancestor), self.convert_selector_to_json(selector.descendant)]
        
    @classmethod
    def get_selector_from_json(self, selector_list):
        if selector_list[0] == "tag":
            selector = TagSelector(selector_list[1])
            selector.priority = selector_list[2]
            return selector
        elif selector_list[0] == "descendant":
            return DescendantSelector(self.get_selector_from_json(selector_list[1]), self.get_selector_from_json(selector_list[2]))

    @classmethod
    def to_json(self, rules_list: list[tuple[TagSelector | DescendantSelector, dict[str, str]]]):
        return [[self.convert_selector_to_json(rule[0]), rule[1]] for rule in rules_list]

    @classmethod
    def from_json(self, rules_list):
        return [(self.get_selector_from_json(rule[0]), rule[1]) for rule in rules_list]

def style(node, rules):
    node.style = {}

    for property, default_value in INHERITED_PROPERTIES.items():
        if node.parent:
            node.style[property] = node.parent.style[property]
        else:
            node.style[property] = default_value

    for selector, body in rules:
        if not selector.matches(node): continue
        for property, value in body.items():
            node.style[property] = value

    if isinstance(node, Element) and "style" in node.attributes:
        pairs = CSSParser(node.attributes["style"]).body()
        for property, value in pairs.items():
            node.style[property] = value 

    if node.style["font-size"].endswith("%"):
        if node.parent:
            parent_font_size = node.parent.style["font-size"]
        else:
            parent_font_size = INHERITED_PROPERTIES["font-size"]

        node_pct = float(node.style["font-size"][:-1]) / 100
        parent_px = float(parent_font_size[:-2])
        node.style["font-size"] = str(node_pct * parent_px) + "px"

    for child in node.children:
        style(child, rules)

def tree_to_list(tree, list):
    list.append(tree)
    for child in tree.children:
        tree_to_list(child, list)
    return list

def replace_symbols(text):
    for key, value in html.entities.html5.items():
        text = text.replace(f"&{key};", value)

    return text

File: src/http_client/html_parser.rs
use crate::constants::*;
use std::collections::HashMap;

#[derive(Clone)]
pub struct Element {
    pub tag: String,
    pub children: Vec<Node>,
    pub attributes: HashMap<String, String>,
    pub parent: Option<Box<Node>>
}

#[derive(Clone)]
pub struct Text {
    pub text: String,
    pub children: Vec<Node>,
    pub parent: Option<Box<Node>>
}

#[derive(Clone)]
pub enum Node {
    Element(Element),
    Text(Text),
}

impl Node {
    pub fn tag(&self) -> Option<&str> {
        match self {
            Node::Element(e) => Some(&e.tag),
            Node::Text(_) => None
        }
    }

    pub fn children(&self) -> &Vec<Node> {
        match self {
            Node::Element(e) => &e.children,
            Node::Text(t) => &t.children
        }
    }

    pub fn parent(&self) -> Option<&Node> {
        match self {
            Node::Text(t) => t.parent.as_deref(),
            Node::Element(e) => e.parent.as_deref(),
        }
    }
}

pub struct HTML {
    pub raw_html: String,
    pub unfinished: Vec<Element>,
}

impl HTML {
    pub fn new(raw_html: String) -> Self {
        HTML {
            raw_html,
            unfinished: Vec::new(),
        }
    }

    pub fn parse(&mut self) -> Element {
        let mut text = String::new();
        let mut in_tag = false;
        let html_content: Vec<char> = self.raw_html.chars().collect();

        for c in html_content {
            if c == '<' {
                in_tag = true;
                
                if (self.unfinished.is_empty() || self.unfinished.last().unwrap().tag != "style") && !text.is_empty() {
                    self.add_text(&text);
                }

                text.clear();
            } else if c == '>' {
                in_tag = false;
                self.add_tag(&text);
                text.clear();
            } else {
                text.push(c);
            }
        }
        
        if !in_tag && !text.is_empty() {
            self.add_text(&text);
        }

        self.finish()
    }

    fn add_text(&mut self, text: &str) {
        let trimmed = text.trim();
        if trimmed.is_empty() {
            return;
        }

        self.implicit_tags(None);
        
        if let Some(parent) = self.unfinished.last_mut() {
            let node = Node::Text(Text {
                text: text.to_string(),
                parent: Some(Box::new(Node::Element(parent.clone()))),
                children: Vec::new()
            });
            parent.children.push(node);
        }
    }

    fn get_attributes(&self, text: &str) -> (String, HashMap<String, String>) {
        let parts: Vec<&str> = text.split_whitespace().collect();
        if parts.is_empty() {
            return (String::new(), HashMap::new());
        }

        let tag = parts[0].to_lowercase();
        let mut attributes: HashMap<String, String> = HashMap::new();
        
        for attrpair in &parts[1..] {
            if attrpair.contains('=') {
                let mut split = attrpair.splitn(2, '=');
                let key = split.next().unwrap_or("");
                let mut value = split.next().unwrap_or("");
                
                if value.len() > 2 && (value.starts_with('\'') || value.starts_with('"')) {
                    value = &value[1..value.len() - 1];
                }

                attributes.insert(key.to_lowercase(), value.to_string());
            } else {
                attributes.insert(attrpair.to_lowercase(), String::new());
            }
        }

        (tag, attributes)
    }

    fn add_tag(&mut self, tag: &str) {
        let (tag_name, attributes) = self.get_attributes(tag);
        
        if tag_name.starts_with('!') {
            return;
        }

        self.implicit_tags(Some(&tag_name));

        if tag_name.starts_with('/') {
            if self.unfinished.len() == 1 {
                return;
            }

            let node = self.unfinished.pop().unwrap();
            if let Some(parent) = self.unfinished.last_mut() {
                parent.children.push(Node::Element(node));
            }
        } else if SELF_CLOSING_TAGS.contains(&tag_name.as_str()) {
            let parent_node = self.unfinished.last().map(|p| Box::new(Node::Element(p.clone())));
            if let Some(parent) = self.unfinished.last_mut() {
                let node = Element {
                    tag: tag_name,
                    attributes,
                    children: Vec::new(),
                    parent: parent_node,
                };
                parent.children.push(Node::Element(node));
            }
        } else {
            let node = Element {
                tag: tag_name,
                attributes,
                children: Vec::new(),
                parent: self.unfinished.last().map(|p| Box::new(Node::Element(p.clone()))),
            };
            
            self.unfinished.push(node);
        }
    }

    fn implicit_tags(&mut self, tag: Option<&str>) {
        loop {
            let open_tags: Vec<String> = self.unfinished.iter().map(|node| node.tag.clone()).collect();

            if open_tags.is_empty() && tag != Some("html") {
                self.add_tag("html");
            } else if open_tags == vec!["html"] && !matches!(tag, Some("head") | Some("body") | Some("/html")) {
                if let Some(tag_str) = tag {
                    if HEAD_TAGS.contains(&tag_str) {
                        self.add_tag("head");
                    } else {
                        self.add_tag("body");
                    }
                } else {
                    self.add_tag("body");
                }
            } else if open_tags == vec!["html", "head"] && !matches!(tag, Some(t) if HEAD_TAGS_EXTRA.contains(&t)) {
                self.add_tag("/head");
            } else {
                break;
            }
        }
    }

    fn finish(&mut self) -> Element {
        if self.unfinished.is_empty() {
            self.implicit_tags(None);
        }
    
        while self.unfinished.len() > 1 {
            let node = self.unfinished.pop().unwrap();
            if let Some(parent) = self.unfinished.last_mut() {
                parent.children.push(Node::Element(node));
            }
        }

        self.unfinished.pop().unwrap()
    }
}

pub struct TagSelector {
    pub tag: String,
    pub priority: i32
}

impl TagSelector {
    pub fn matches(&self, node: &Node) -> bool{
        if let Node::Element(elem) = node {
            self.tag == elem.tag
        } else {
            false
        }
    }

    pub fn new(tag: String) -> TagSelector {
        TagSelector {
            tag,
            priority: 1
        }
    }
}
    
pub struct DescendantSelector {
    pub ancestor: Box<Rule>,
    pub descendant: Box<Rule>,
    pub priority: i32
}

impl DescendantSelector {
    pub fn new(ancestor: Rule, descendant: Rule) -> DescendantSelector {
        let new_priority = ancestor.priority() + descendant.priority();
        DescendantSelector {
            ancestor: Box::new(ancestor),
            descendant: Box::new(descendant),
            priority: new_priority
        }
    }

    pub fn matches(&self, node: &Node) -> bool {
        if !self.descendant.matches(node) {
            return false;
        }

        let mut current = node;

        loop {
            match current.parent() {
                Some(parent) => {
                    if self.ancestor.matches(parent) {
                        return true;
                    } 

                    current = parent;
                },
                None => break
            }
        }

        return false;
    }   
}

pub enum Rule {
    TagSelector(TagSelector),
    DescendantSelector(DescendantSelector),
}

impl Rule {
    pub fn priority(&self) -> i32 {
        match self {
            Rule::TagSelector(s) => s.priority,
            Rule::DescendantSelector(s) => s.priority,
        }
    }
    pub fn matches(&self, node: &Node) -> bool {
        match self {
            Rule::TagSelector(s) => s.matches(node),
            Rule::DescendantSelector(s) => s.matches(node),
        }
    }
}

pub fn cascade_priority(rule: (Rule, HashMap<String, String>)) -> i32 {
    rule.0.priority()
}

pub fn get_inline_styles(node: &Node) -> Vec<(Rule, HashMap<String, String>)> {
    let mut all_rules = vec![];

    if let Node::Element(elem) = node {
        for child in &elem.children {
            if let Node::Element(child_elem) = child {
                if child_elem.tag == "style" {
                    if let Some(Node::Text(text_node)) = child_elem.children.first() {
                        all_rules.extend(CSSParser::new(text_node.text.clone()).parse());
                    }
                }
            }
            all_rules.extend(get_inline_styles(child));
        }
    }

    all_rules
}

pub struct CSSParser {
    chars: Vec<char>,
    len: usize,
    i: usize
}

impl CSSParser {
    fn new (s: String) -> CSSParser {
        CSSParser {
            chars: s.chars().collect(),
            len: s.chars().count(),
            i: 0
        }
    }
    
    fn whitespace(&mut self) {
        while self.i < self.len && self.chars[self.i].is_whitespace() {
            self.i += 1;
        }
    }
        
    fn literal(&mut self, literal: char) -> Result<(), String> {
        if !(self.i < self.len && self.chars[self.i] == literal) {
            return Err(format!("Expected '{}'", literal));
        }

        self.i += 1;
        
        Ok(())
    }

    fn word(&mut self) -> Result<String, String> {
        let start = self.i;
        
        while self.i < self.len {
            if self.chars[self.i].is_alphanumeric() || "#-.%".contains(self.chars[self.i]) {
                self.i += 1;
            }

            else {
                break
            }
        }

        if !(self.i > start) {
            return Err("Parsing error: unexpected word".to_string())
        };

        Ok(self.chars[start..self.i].iter().collect())
    }
        
    fn pair(&mut self) -> Result<(String, String), String> {
        let prop = self.word()?;
        
        self.whitespace();
        self.literal(':')?;
        self.whitespace();
        
        let val = self.word()?;

        Ok((prop.to_lowercase(), val))
    }

    fn ignore_until(&mut self, chars: Vec<char>) -> Option<char> {
        while self.i < self.len {
            let c = self.chars[self.i];
            if chars.contains(&c) {
                return Some(c);
            }
            else {
                self.i += 1;
            }
        }

        return None;
    }

    fn body(&mut self) -> HashMap<String, String> {
        let mut pairs = HashMap::new();

        while self.i < self.len && self.chars[self.i] != '}' {
            match self.pair() {
                Ok((prop, val)) => {
                    pairs.insert(prop, val);
                    self.whitespace();
                    let _ = self.literal(';');
                    self.whitespace();
                }
                Err(_) => {
                    let ignore_char: Vec<char> = vec![';', '}'];
                    if let Some(';') = self.ignore_until(ignore_char) {
                        let _ = self.literal(';');
                        self.whitespace();
                    } else {
                        break;
                    }
                }
            }
        }

        pairs
    }
    
    fn selector(&mut self) -> Result<Rule, String> {
        let mut out = Rule::TagSelector(TagSelector::new(self.word()?.to_lowercase()));
        
        self.whitespace();
        
        while self.i < self.len && self.chars[self.i] != '{' {
            let tag = self.word()?;
            let descendant = Rule::TagSelector(TagSelector::new(tag.to_lowercase()));
            out = Rule::DescendantSelector(DescendantSelector::new(out, descendant));
            self.whitespace();
        }
        
        Ok(out)
    }

    pub fn parse(&mut self) -> Vec<(Rule, HashMap<String, String>)> {
        let mut rules = vec![];
        while self.i < self.len {
            self.whitespace();
            
            let selector = match self.selector() {
                Ok(s) => s,
                Err(_) => {
                    if let Some('}') = self.ignore_until(vec!['}']) {
                        let _ = self.literal('}');
                        self.whitespace();
                    } else {
                        break;
                    }
                    continue;
                }
            };
            
            if self.literal('{').is_err() {
                if let Some('}') = self.ignore_until(vec!['}']) {
                    let _ = self.literal('}');
                    self.whitespace();
                } else {
                    break;
                }
                continue;
            };
            
            self.whitespace();
            
            let body = self.body();
            
            let _ = self.literal('}');

            rules.push((selector, body));
        
        }

        return rules;
    }

}
